metadata/src/main/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectory.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.metadata.bootstrap;
+
+import org.apache.kafka.metadata.util.BatchFileReader;
+import org.apache.kafka.metadata.util.BatchFileReader.BatchAndType;
+import org.apache.kafka.metadata.util.BatchFileWriter;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.server.common.MetadataVersion;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.apache.kafka.server.common.MetadataVersion.MINIMUM_KRAFT_VERSION;
+
+
+/**
+ * A read-only class that holds the controller bootstrap metadata. A file named "bootstrap.snapshot" is used and the
+ * format is the same as a KRaft snapshot.
+ */
+public class BootstrapDirectory {
+    final static String INTER_BROKER_PROTOCOL_CONFIG_KEY = "inter.broker.protocol.version";
+    final static String BINARY_BOOTSTRAP = "bootstrap.checkpoint";
+
+    public static String ibpStringFromConfigMap(Map<String, Object> staticConfig) {
+        Object value = staticConfig.get(INTER_BROKER_PROTOCOL_CONFIG_KEY);
+        return value == null ? "" : value.toString();
+    }
+
+    private final String directoryPath;
+    private final String ibp;
+
+    /**
+     * Create a new BootstrapDirectory object.
+     *
+     * @param directoryPath     The path to the directory with the bootstrap file.
+     * @param ibp               The configured value of inter.broker.protocol, or the empty string
```

Comment:
I don't feel that strongly about it. But generally making a parameter non-nullable feels more natural than making it non-emptyable, and you want to choose one or the other

Reply:
(I mean, I guess another point of view is that empty string is just another invalid input, that we throw an exception on)

As I said, I don't feel that strongly. If you think it should be changed I can do that

Reply:
I don't feel strongly either. The main argument in favor of `null` is that it prevents the value from being used in a context where it might not apply. In other words, it provides some runtime safety. A better option is probably `Optional<String>`, which also provides compile time safety.

Reply:
I will just make this an `Optional<MetadataVersion>` for simplicity.

Reply:
let's leave this as-is for now and we can refactor later